main.tf

```
@@ -1,65 +1,88 @@
+  description                       = each.value.description
+  ipam_scope_id                     = each.value.ipam_scope_id
+  #could be output of aws_vpc_ipam or aws_vpc_ipam_scope - try statement?
+  locale                = try(lower(each.value.locale), null)
+  publicly_advertisable = each.value.publicly_advertisable
+  public_ip_source      = each.value.public_ip_source
+  source_ipam_pool_id   = each.value.source_ipam_pool_id
+  tags                  = each.value.tags
 }
 
-module "vpc_ipam_pool_cidr_allocation" {
-  source           = "./modules/pool-cidr-allocation"
-  cidr             = var.pool_cidr_allocation.cidr
-  disallowed_cidrs = var.pool_cidr_allocation.disallowed_cidrs
-  ipam_pool_id     = var.pool_cidr_allocation.ipam_pool_id
-  netmask_length   = var.pool_cidr_allocation.netmask_length
+
+#plural, provisions a CIDR from an IPAM pool
+#netmask conflisks with cidr and for_each statement
+resource "aws_vpc_ipam_pool_cidr" "this" {
+  for_each = { for cidr in var.vpc_ipam.pool_cidrs : cidr.cidr => cidr }
+
+
+  cidr = each.value.cidr
+  cidr_authorization_context {
+    message   = each.value.cidr_authorization_context_message
+    signature = each.value.cidr_authorization_context_signature
+  }
+  ipam_pool_id   = each.value.ipam_pool_id
+  netmask_length = each.value.netmask_length
 }
 
-module "vpc_ipam_resource_discovery" {
-  source      = "./modules/resource-discovery"
-  region_name = var.resource_discovery.region_name
-  tags        = var.resource_discovery.tags
+#plural, carves out a CIDR allocation from pool, preventing usage
+resource "aws_vpc_ipam_pool_cidr_allocation" "this" {
+  for_each = { for allocation in var.vpc_ipam.pool_cidr_allocations : allocation.cidr => allocation }
+
+  cidr             = each.value.cidr
+  description      = each.value.description
+  disallowed_cidrs = each.value.disallowed_cidrs
+  ipam_pool_id     = each.value.ipam_pool_id
+  netmask_length   = each.value.netmask_length
 }
 
-module "vpc_ipam_resource_discovery_association" {
-  source                     = "./modules/resource-discovery-association"
-  ipam_id                    = var.resource_discovery_association.ipam_id
-  ipam_resource_discovery_id = var.resource_discovery_association.ipam_resource_discovery_id
-  tags                       = var.resource_discovery_association.tags
+#for checking next available CIDR,
+# unsure of implementation usage? could be useful for checking/assigning next CIDR (dynamically)?
+resource "aws_vpc_ipam_preview_next_cidr" "this" {
+  count = local.create_vpc_ipam_preview_next_cidr && var.vpc_ipam.preview_next_cidr != null ? 1 : 0
+
+  #disallowed cidrs could be a dynamic input
+  disallowed_cidrs = var.vpc_ipam.preview_next_cidr.disallowed_cidrs != null ? var.vpc_ipam.preview_next_cidr.disallowed_cidrs : []
+  ipam_pool_id     = try(var.vpc_ipam.preview_next_cidr.ipam_pool_id, [])
+  netmask_length   = try(var.vpc_ipam.preview_next_cidr.netmask_length, [])
```

`ipam_pool_id` is required, so the try usage isn't making sense to me. The others are optional, and the resource should handle the null default just fine. Curious what provoked this setup...?

```suggestion
  disallowed_cidrs = var.vpc_ipam.preview_next_cidr.disallowed_cidrs
  ipam_pool_id     = var.vpc_ipam.preview_next_cidr.ipam_pool_id
  netmask_length   = var.vpc_ipam.preview_next_cidr.netmask_length
```